File: drawFuncsV2.py
import matplotlib.pyplot as plt
from geoThings import Circle, InfLine, VertLine, fsqroot, randomList, printCircleData
from shapely import Point, Polygon, LineString

def main():
    setup(12,9)
    circ1 = Circle(Point(1,1),2)
    circ2 = Circle(Point(-2,2),3)
    drawCircles([circ1])
    pdraw(1)
    drawCircles([circ1,circ2])
    pdraw(2)

#Drawing
#New ones
def setup(width = 12, height = 9):
    fig, ax = plt.subplots()
    fig.set_figwidth(width)
    fig.set_figheight(height)

def pshow(isBlocking = True):
    ax = plt.gca()

    ax.grid(True, which='both')
    ax.axhline(y=0, color='k')
    ax.axvline(x=0, color='k')

    ax.set_xlim(-10, 10)
    ax.set_ylim(-10, 10)
    plt.axis('equal')  # To ensure the aspect ratio is maintained

    plt.show(block =  isBlocking)

def pdraw(time):
    fig, ax = plt.gcf(), plt.gca()

    ax.grid(True, which='both')
    ax.axhline(y=0, color='k')
    ax.axvline(x=0, color='k')

    ax.set_xlim(-10, 10)
    ax.set_ylim(-10, 10)
    plt.axis('equal')  # To ensure the aspect ratio is maintained

    plt.draw()
    plt.pause(time)
    plt.cla()
    plt.clf()

def drawPoints(pointList, pcolor = 'red', pmarker = 'o'): #list of Points
    xs = [p.x for p in pointList]
    ys = [p.y for p in pointList]
    plots = []
    for i in range(len(pointList)):
        plotPoint = plt.plot(xs[i],ys[i],color=pcolor,marker=pmarker)
        plots.append(plotPoint)
    return plots

def removePoints(pointList_p):
    for point_p in pointList_p:
        point_p[0].remove()

def drawCircles(circleList): #Circles
    fig = plt.gcf()
    ax = plt.gca()
    plots = []
    for index, circ in enumerate(circleList):
        #plots the circle
        plotCirc = plt.Circle(circ.tupCenter(), circ.radius, fill = False)
        ax.add_patch(plotCirc)

        #plots the center, labeled by index in circleList
        cx, cy = circ.tupCenter()
        plotCenter = plt.plot(cx,cy, color = 'b', marker = 'o')
        plotCenterAnn = plt.annotate(str(index+1),(cx,cy), textcoords = 'offset points', xytext = (7,7))

        plots.append([plotCirc,plotCenter,plotCenterAnn])

    return plots

def removeCircles(circleList_p): #usually an element of the thing returned by drawCircles
    for circle_p in circleList_p:
        circle_p[0].remove()
        circle_p[1][0].remove()
        circle_p[2].remove()


def drawLines(lineList): #InfLines or VertLines
    fig, ax = plt.gcf(), plt.gca()

    plots = []
    for line in lineList:
        if type(line) is VertLine:
            plotLine = plt.axvline(line.xInt)
        else:
            plotLine = plt.axline(
                (0,line.b),
                slope=line.m
                )
    return plots

def drawPolygon(polygon, polyColor = 'red'): #Polygon, string
    fig, ax = plt.gcf(), plt.gca()

    coords = list(polygon.exterior.coords) #for plotPoly
    coordPoints = [Point(p) for p in list(polygon.exterior.coords)] #for drawPoints

    plotVertices = drawPoints(coordPoints, polyColor, 'o')

    plotPoly = plt.Polygon(coords, color = polyColor, lw = 2, fill = False)
    ax.add_patch(plotPoly)

    return [plotPoly, plotVertices]

def drawSegments(segments):
    plots = []
    for seg in segments:
        x0 = list(seg.coords)[0][0]
        x1 = list(seg.coords)[1][0]
        y0 = list(seg.coords)[0][1]
        y1 = list(seg.coords)[1][1]
        seg_p = plt.plot([x0,x1],[y0,y1],color='C' + str(segments.index(seg)), marker='1')
        plots.append(seg_p)
    return plots


def drawBox(bounds):
    fig, ax = plt.gcf(), plt.gca()
    minx, miny, maxx, maxy = bounds

    rect = plt.Rectangle((minx,miny), maxx-minx, maxy-miny, fill = False)
    ax.add_patch(rect)

if __name__ == '__main__':
    main()

File: drawFuncsV2.R

library(ggplot2)
library(grid)
library(gridExtra)

# Import equivalent R packages/modules
# Assuming geoThings R equivalent exists with Circle, VertLine, fsqroot, randomList, printCircleData
source("geoThings.R")  # Equivalent of: from geoThings import Circle, VertLine, fsqroot, randomList, printCircleData
# Assuming Point equivalent exists
if (!exists("Point")) {
  Point <- function(x, y) {
    structure(list(x = x, y = y), class = "Point")
  }
}

# Global variables to store current plot and plot objects
current_plot <- NULL
current_fig_width <- 12
current_fig_height <- 9
plot_objects <- list()

# this file contains all the functions for drawing diagrams with matplotlib. The usual way to use is
#     setup()
#     functions that draw things
#     pshow()

main <- function() {
  setup(12, 9)
  circ1 <- Circle(Point(1, 1), 2)
  circ2 <- Circle(Point(-2, 2), 3)
  drawCircles(list(circ1))
  pdraw(1)
  drawCircles(list(circ1, circ2))
  pdraw(2)
}

# Drawing
# New ones
setup <- function(width = 12, height = 9) {
  current_fig_width <<- width
  current_fig_height <<- height
  
  # Create base plot
  current_plot <<- ggplot() + 
    theme_minimal() +
    coord_fixed(ratio = 1) +
    xlim(-10, 10) +
    ylim(-10, 10)
  
  # Clear any existing plot objects
  plot_objects <<- list()
}

pshow <- function(isBlocking = TRUE) {
  # Add grid, axes
  final_plot <- current_plot +
    geom_hline(yintercept = 0, color = "black") +
    geom_vline(xintercept = 0, color = "black") +
    theme(panel.grid = element_line(color = "gray90"),
          panel.grid.major = element_line(color = "gray80"),
          panel.grid.minor = element_line(color = "gray90")) +
    coord_fixed(ratio = 1, xlim = c(-10, 10), ylim = c(-10, 10))
  
  # Display the plot
  print(final_plot)
  
  if (isBlocking) {
    # Wait for user input to continue
    cat("Press [Enter] to continue...")
    invisible(readLines(n = 1))
  }
}

pdraw <- function(time) {
  # Add grid, axes
  final_plot <- current_plot +
    geom_hline(yintercept = 0, color = "black") +
    geom_vline(xintercept = 0, color = "black") +
    theme(panel.grid = element_line(color = "gray90"),
          panel.grid.major = element_line(color = "gray80"),
          panel.grid.minor = element_line(color = "gray90")) +
    coord_fixed(ratio = 1, xlim = c(-10, 10), ylim = c(-10, 10))
  
  # Display the plot
  print(final_plot)
  
  # Pause for specified time
  Sys.sleep(time)
  
  # Clear plot (equivalent to plt.cla() and plt.clf())
  current_plot <<- ggplot() + 
    theme_minimal() +
    coord_fixed(ratio = 1) +
    xlim(-10, 10) +
    ylim(-10, 10)
  
  plot_objects <<- list()
}

drawPoints <- function(pointList, pcolor = 'red', pmarker = 'o') { # list of Points
  xs <- sapply(pointList, function(p) p$x)
  ys <- sapply(pointList, function(p) p$y)
  
  plots <- list()
  
  for (i in 1:length(pointList)) {
    # Add point to current plot
    current_plot <<- current_plot + 
      geom_point(data = data.frame(x = xs[i], y = ys[i]), 
                 aes(x = x, y = y), 
                 color = pcolor, 
                 shape = ifelse(pmarker == 'o', 16, 16),
                 size = 3)
    
    # Store plot reference
    plot_ref <- list(type = "point", x = xs[i], y = ys[i], color = pcolor)
    plots[[i]] <- plot_ref
  }
  
  return(plots)
}

removePoints <- function(pointList_p) {
  # In R/ggplot2, we need to rebuild the plot without these points
  # This is a simplified approach - in practice, would need more sophisticated layer management
  for (point_p in pointList_p) {
    # Remove from plot_objects if tracking
    # Note: ggplot2 doesn't have direct remove functionality like matplotlib
    warning("Point removal not fully implemented - would require plot reconstruction")
  }
}

drawCircles <- function(circleList) { # Circles
  plots <- list()
  
  for (index in 1:length(circleList)) {
    circ <- circleList[[index]]
    
    # Create circle data points
    theta <- seq(0, 2*pi, length.out = 100)
    center <- circ$tupCenter()  # Assuming this method exists
    circle_x <- center[1] + circ$radius * cos(theta)
    circle_y <- center[2] + circ$radius * sin(theta)
    
    # Add circle to plot
    current_plot <<- current_plot + 
      geom_path(data = data.frame(x = circle_x, y = circle_y), 
                aes(x = x, y = y), 
                color = "black")
    
    # Add center point
    cx <- center[1]
    cy <- center[2]
    current_plot <<- current_plot + 
      geom_point(data = data.frame(x = cx, y = cy), 
                 aes(x = x, y = y), 
                 color = "blue", 
                 size = 3)
    
    # Add annotation (label)
    current_plot <<- current_plot + 
      geom_text(data = data.frame(x = cx + 0.3, y = cy + 0.3, label = as.character(index)), 
                aes(x = x, y = y, label = label), 
                size = 3)
    
    # Store plot references
    plot_ref <- list(
      type = "circle",
      center = c(cx, cy),
      radius = circ$radius,
      index = index
    )
    plots[[index]] <- plot_ref
  }
  
  return(plots)
}

removeCircles <- function(circleList_p) { # usually an element of the thing returned by drawCircles
  # Similar to removePoints - would need plot reconstruction in ggplot2
  for (circle_p in circleList_p) {
    warning("Circle removal not fully implemented - would require plot reconstruction")
  }
}

drawLines <- function(lineList) { # InfLines or VertLines
  plots <- list()
  
  for (i in 1:length(lineList)) {
    line <- lineList[[i]]
    
    if (class(line)[1] == "VertLine") {
      # Vertical line
      current_plot <<- current_plot + 
        geom_vline(xintercept = line$xInt, color = "black")
    } else {
      # Line with slope
      # Create line data points across plot range
      x_vals <- seq(-10, 10, length.out = 100)
      y_vals <- line$m * x_vals + line$b
      
      current_plot <<- current_plot + 
        geom_line(data = data.frame(x = x_vals, y = y_vals), 
                  aes(x = x, y = y), 
                  color = "black")
    }
    
    plot_ref <- list(type = "line", line_obj = line)
    plots[[i]] <- plot_ref
  }
  
  return(plots)
}

drawPolygon <- function(polygon, polyColor = 'red') { # Polygon, string
  # Extract coordinates from polygon exterior
  coords <- polygon$exterior$coords  # Assuming shapely-like structure
  
  # Create points for vertices
  coordPoints <- lapply(coords, function(p) Point(p[1], p[2]))
  
  # Draw vertices
  plotVertices <- drawPoints(coordPoints, polyColor, 'o')
  
  # Draw polygon outline
  poly_df <- data.frame(
    x = sapply(coords, function(p) p[1]),
    y = sapply(coords, function(p) p[2])
  )
  
  current_plot <<- current_plot + 
    geom_polygon(data = poly_df, 
                 aes(x = x, y = y), 
                 color = polyColor, 
                 fill = NA, 
                 size = 2)
  
  plot_ref <- list(type = "polygon", vertices = plotVertices)
  return(plot_ref)
}

drawSegments <- function(segments) {
  plots <- list()
  
  for (i in 1:length(segments)) {
    seg <- segments[[i]]
    
    # Extract coordinates
    coords <- seg$coords  # Assuming shapely-like structure
    x0 <- coords[[1]][1]
    x1 <- coords[[2]][1]
    y0 <- coords[[1]][2]
    y1 <- coords[[2]][2]
    
    # Generate color (equivalent to 'C' + str(index))
    color_index <- ((i - 1) %% 10) + 1
    colors <- c("blue", "orange", "green", "red", "purple", "brown", "pink", "gray", "olive", "cyan")
    seg_color <- colors[color_index]
    
    # Draw segment
    current_plot <<- current_plot + 
      geom_segment(data = data.frame(x = x0, y = y0, xend = x1, yend = y1), 
                   aes(x = x, y = y, xend = xend, yend = yend), 
                   color = seg_color, 
                   size = 1)
    
    plot_ref <- list(type = "segment", coords = list(c(x0, y0), c(x1, y1)), color = seg_color)
    plots[[i]] <- plot_ref
  }
  
  return(plots)
}

drawBox <- function(bounds) {
  minx <- bounds[1]
  miny <- bounds[2]
  maxx <- bounds[3]
  maxy <- bounds[4]
  
  # Create rectangle
  rect_df <- data.frame(
    x = c(minx, maxx, maxx, minx, minx),
    y = c(miny, miny, maxy, maxy, miny)
  )
  
  current_plot <<- current_plot + 
    geom_path(data = rect_df, 
              aes(x = x, y = y), 
              color = "black")
  
  return(list(type = "box", bounds = bounds))
}

# Run main if this file is executed directly
if (sys.nframe() == 0) {
  main()
}

